pkg/endpoint/bpf.go

```diff
@@ -1407,37 +1407,65 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
 		e.PolicyDebug(logrus.Fields{"dumpedDiffs": diffs}, "syncPolicyMapWithDump")
 	}
 
+	if err == nil {
+		e.policyMapLastFullReconciliation = time.Now()
+	}
+
 	return err
 }
 
-func (e *Endpoint) startSyncPolicyMapController() {
-	// Skip the controller if the endpoint has no policy map
-	if !e.HasBPFPolicyMap() {
-		return
+func (e *Endpoint) syncPolicyMapControllerName() string {
+	return fmt.Sprintf("sync-policymap-%d", e.ID)
+}
+
+// updateSyncPolicyMapController creates or triggers a controller to  reconcile
+// pending policy map changes. Every PolicyMapFullReconciliationInterval the
+// controller also does full reconciliation by dumping the policy map.
+func (e *Endpoint) updateSyncPolicyMapController() channels.DoneChan {
+	// Skip the controller if the endpoint has no policy map or is
+	// being disconnected.
+	if !e.HasBPFPolicyMap() || e.IsDisconnecting() {
+		return channels.ClosedDoneChan
 	}
 
-	ctrlName := fmt.Sprintf("sync-policymap-%d", e.ID)
-	e.controllers.CreateController(ctrlName,
+	return e.controllers.UpdateController(
```

Comment:
We should use TriggerController() here, not UpdateController()  (except on initialization of course). UpdateController can miss updates.

Reply:
This will require a bit of state and swapping channels. It won't be pretty :-(.

Reply:
https://github.com/cilium/cilium/issues/29015 -- UpdateController is not reliable.